nevergrad/functions/gym/gym_multi.py

```diff
@@ -0,0 +1,780 @@
+            previous_o = np.asarray(o)
+            o = np.concatenate([previous_o.ravel(), memory.ravel(), self.extended_input])
+            assert (
+                len(o) == self.input_dim
+            ), f"o has shape {o.shape} whereas input_dim={self.input_dim} ({control} / {env} {self.name} (limited={self.limited_compiler_gym}))"
+            a, memory = self.neural(x[i % len(x)] if "multi" in control else x, o)
+            a = self.action_cast(a)
+            try:
+                o, r, done, _ = self.step(a)  # Outputs = observation, reward, done, info.
+                current_time_index += 1
+                if "multifidLANM" in self.name and current_time_index > 500 and limited_fidelity:
+                    done = True
+                current_reward *= self.gamma
+                current_reward += r
+                current_observations += [np.asarray(o).copy()]
+                current_actions += [np.asarray(a).copy()]
+                if (
+                    done and "stacking" in self.control
+                ):  # Only the method which do a stacking of heuristic + memory into the
+                    # observation need archiving.
+                    self.num_losses += 1
+                    tau = 1.0 / self.num_losses
+                    self.mean_loss = (
+                        ((1.0 - tau) * self.mean_loss + tau * current_reward)
+                        if self.mean_loss is not None
+                        else current_reward
+                    )
+                    found = False
+                    for trace in self.archive:
+                        to, _, _ = trace
+                        if np.array_equal(
+                            np.asarray(current_observations, dtype=np.float32),
+                            np.asarray(to, dtype=np.float32),
+                        ):
+                            found = True
+                            break
+                    if not found:
+                        # Risky: this code assumes that the object is used only in a single run.
+                        self.archive += [(current_observations, current_actions, current_reward)]
+            except AssertionError:  # Illegal action.
+                return 1e20 / (1.0 + i)  # We encourage late failures rather than early failures.
+            if "stacking" in control:
+                attention_a = self.heuristic(
+                    o, current_observations
+                )  # Best so far, or something like that heuristically derived.
+                a = attention_a if attention_a is not None else 0.0 * np.asarray(a)
+                previous_o = previous_o.ravel()
+                additional_input = np.concatenate([np.asarray(a).ravel(), previous_o])
+                shift = len(additional_input)
+                self.extended_input[: (len(self.extended_input) - shift)] = self.extended_input[shift:]
+                self.extended_input[(len(self.extended_input) - shift) :] = additional_input
```

Comment:
`heuristic` is used only here and its goal is to modify extended_input. Should we change the scope of heuristic to do everything selected here in order to simplify gym_simulate?